### PR TITLE
✨introduce test E2E

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR introduces an E2e folder that introduces clusterctl based E2E test for quick-start. self-hosted, KCP upgrade.

Most notably:
- Such test can be executed with *any* infrastructure provider (not only CAPD)
- Test spec are exposed as a library so they can be reused by other projects

**Which issue(s) this PR fixes**:
xref #2753
xref #2637
xref #2636

/assign @vincepri
/assign @sedefsavas

### CODE REVIEW

File: test/e2e/data/infrastructure-aws/cluster-template.yaml [hunk 1]
```diff
@@ -0,0 +1,113 @@
+---
```
Comment (nader-ziada) [inline test/e2e/data/infrastructure-aws/cluster-template.yaml:1]: if we already rely on having the `aws-provider` repo existing to generate the images, can we use the template already there?

File: test/e2e/e2e_suite_test.go [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+	managementCluster, managementClusterKubeConfigPath = clusterctl.InitManagementCluster(context.TODO(), &clusterctl.InitManagementClusterInput{
+		E2EConfig:            e2eConfig,
+		ClusterctlConfigPath: clusterctlConfigPath,
+		Scheme:               scheme,
+		NewManagementClusterFn: func(name string, scheme *runtime.Scheme) (cluster framework.ManagementCluster, kubeConfigPath string, err error) {
+			cluster, err = capde2e.NewClusterForCAPD(context.TODO(), name, scheme)
+			if err != nil {
+				return nil, "", err
+			}
+			//TODO: this is hacky way to get the kubeconfig path. As soon as v1alpha4 opens, we should add GetKubeConfigPath method to the ManagementCluster interface and get rid of this
```
Comment (nader-ziada) [inline test/e2e/e2e_suite_test.go:153]: is adding a new method to the framework breaking the contract of v1alpha3?

File: test/e2e/quick_start_test.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+		}, e2eConfig.IntervalsOrDefault("quick-start/wait-provision-cluster", "3m", "10s")...)
+
+		By("Waiting for the first control plane machine to be provisioned")
+		controlPlane := discovery.GetKubeadmControlPlaneByCluster(context.TODO(), discovery.GetKubeadmControlPlaneByClusterInput{
+			Lister:      client,
+			ClusterName: clusterName,
+			Namespace:   namespaceName,
+		})
+		Expect(controlPlane).ToNot(BeNil())
+		framework.WaitForOneKubeadmControlPlaneMachineToExist(context.TODO(), framework.WaitForOneKubeadmControlPlaneMachineToExistInput{
```
Comment (nader-ziada) [inline test/e2e/quick_start_test.go:120]: It might be better to wait for all machines to be created before the CNI

File: test/e2e/quick_start_test.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+		By("Installing Calico on the workload cluster")
+		workloadClient, err := managementCluster.GetWorkloadClient(context.TODO(), namespaceName, clusterName)
+		Expect(err).ToNot(HaveOccurred(), "Failed to get the client for the workload cluster with name %s", cluster.Name)
+
+		//TODO: CNI plugin configurable/pluggable (see CAPA)?
+		applyYAMLInput := framework.ApplyYAMLInput{
+			Client:     workloadClient,
+			Scheme:     scheme,
+			Filename:   cniPath,
```
Comment (nader-ziada) [inline test/e2e/quick_start_test.go:134]: different providers could have some minor differences in their Calico file, so maybe make that based on provider

File: test/e2e/data/infrastructure-aws/cluster-template.yaml (hunk 1 above)
Comment (fabriziopandini) [reply test/e2e/data/infrastructure-aws/cluster-template.yaml:1]: It is possible to use files from other repository (if they exist locally) but IMO test should be as isolated as possible, so I prefer to have separate templates

File: test/e2e/e2e_suite_test.go (hunk 2 above)
Comment (fabriziopandini) [reply test/e2e/e2e_suite_test.go]: I saw other PRs adding methods to the same interface, so I guess this is acceptable

File: test/e2e/quick_start_test.go (hunk 3 above)
Comment (fabriziopandini) [reply test/e2e/quick_start_test.go]: Is there a particular reason to do so?
AFAIK installing as soon as the control plane is up is legit; it is also a precondition to get the control plane ready...

File: test/e2e/quick_start_test.go (hunk 4 above)
Comment (fabriziopandini) [reply test/e2e/quick_start_test.go]: Noted. Working ok on this

**Comment by wfernandes** [comment]
/assign
Assigning myself to this because I'd like to get context on this PR and help write other tests.

File: test/e2e/Makefile [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+## Testing
+## --------------------------------------
+
+GINKGO_FOCUS  ?=
+E2E_CONF_FILE  ?= ${REPO_ROOT}/test/e2e/config/docker-dev.conf
+ARTIFACTS_FOLDER ?= ${REPO_ROOT}/_artifacts
+SKIP_RESOURCE_CLEANUP ?= false
+
+.PHONY: run-e2e
+run-e2e: ## Run the end-to-end tests
```
Comment (vincepri) [inline test/e2e/Makefile:44]:
```suggestion
.PHONY: run
run: ## Run the end-to-end tests
```

given that we're already in the e2e folder

File: bootstrap/kubeadm/config/manager/manager_image_patch.yaml [hunk 6]
```diff
@@ -7,5 +7,5 @@ spec:
   template:
     spec:
       containers:
-        - image: gcr.io/k8s-staging-cluster-api/kubeadm-bootstrap-controller:master
+        - image: gcr.io/k8s-staging-cluster-api/kubeadm-bootstrap-controller-amd64:dev
```
Comment (wfernandes) [inline bootstrap/kubeadm/config/manager/manager_image_patch.yaml:10]: Was this change meant to be included?

Comment (fabriziopandini) [reply bootstrap/kubeadm/config/manager/manager_image_patch.yaml:10]: no, that was a mistake, thanks for pointing out

File: test/e2e/Makefile (hunk 5 above)
Comment (fabriziopandini) [reply test/e2e/Makefile]: done

File: test/framework/clusterctl/e2e_config.go [hunk 7]
```diff
@@ -396,32 +394,22 @@ func (c *E2EConfig) GetVariable(varName string) string {
...
-		return Int64Ptr(0)
+		return nil
 	}
 
 	wCount, err := strconv.ParseInt(wCountStr, 10, 64)
 	Expect(err).NotTo(HaveOccurred())
 	return Int64Ptr(wCount)
 }
 
+func Int64Ptr(n int64) *int64 {
```
Comment (vincepri) [inline test/framework/clusterctl/e2e_config.go:409]: https://github.com/kubernetes/utils/blob/master/pointer/pointer.go#L55

File: test/e2e/quick_start_test.go (hunk 4 above)
Comment (fabriziopandini) [reply test/e2e/quick_start_test.go]: solved

File: test/framework/clusterctl/e2e_config.go (hunk 7 above)
Comment (fabriziopandini) [reply test/framework/clusterctl/e2e_config.go]: fixed

File: test/e2e/e2e_suite_test.go (hunk 2 above)
Comment (fabriziopandini) [reply test/e2e/e2e_suite_test.go]: Now there is also the api-diff jobs that cover compatibility issues

**Comment by wfernandes** [comment]
Reviewing

File: test/e2e/kcp_upgrade.go [hunk 8]
```diff
@@ -71,6 +69,8 @@ func KCPUpgradeSpec(ctx context.Context, inputGetter func() KCPUpgradeSpecInput)
 	It("Should successfully upgrade Kubernetes, DNS, kube-proxy, and etcd", func() {
 
 		By("Creating a workload cluster")
+		Expect(input.E2EConfig.Variables).To(HaveKey(clusterctl.KubernetesVersion))
+		Expect(input.E2EConfig.Variables).To(HaveKey(clusterctl.CNIPath))
```
Comment (sedefsavas) [inline test/e2e/kcp_upgrade.go:73]: Do we need these checks for every test? We already throw error if they are empty during validation.

Comment (fabriziopandini) [reply test/e2e/kcp_upgrade.go:73]: I don't think we should enforce variables to exist during validation... I missed this.
Each spec might require different variables

Comment (sedefsavas) [reply test/e2e/kcp_upgrade.go:73]: KubernetesVersion and CNI should exist for all tests? If not we should remove those checks from validation.

**Comment by sedefsavas** [comment]
Changes look good to me. Great work @fabriziopandini.
Over to @wfernandes and @vincepri.

File: test/e2e/self_hosted.go [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+	BeforeEach(func() {
+		Expect(ctx).NotTo(BeNil(), "ctx is required for %s spec", specName)
+		input = inputGetter()
+		Expect(input.E2EConfig).ToNot(BeNil(), "Invalid argument. input.E2EConfig can't be nil when calling %s spec", specName)
+		Expect(input.ClusterctlConfigPath).To(BeAnExistingFile(), "Invalid argument. input.ClusterctlConfigPath must be an existing file when calling %s spec", specName)
+		Expect(input.BootstrapClusterProxy).ToNot(BeNil(), "Invalid argument. input.BootstrapClusterProxy can't be nil when calling %s spec", specName)
+		Expect(os.MkdirAll(input.ArtifactFolder, 0755)).To(Succeed(), "Invalid argument. input.ArtifactFolder can't be created for %s spec", specName)
+
+		// Setup a Namespace where to host objects for this spec and create a watcher for the namespace events.
+		namespace, cancelWatches = setupSpecNamespace(context.TODO(), specName, input.BootstrapClusterProxy, input.ArtifactFolder)
```
Comment (wfernandes) [inline test/e2e/self_hosted.go:70]: Same as above...should we be using the context that is passed in to `SelfHostedSpec`?

File: test/framework/clusterctl/clusterctl_helpers.go [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+	Expect(input.ClusterctlConfigPath).To(BeAnExistingFile(), "Invalid argument. input.ClusterctlConfigPath must be an existing file when calling InitManagementClusterAndWatchControllerLogs")
+	Expect(input.InfrastructureProviders).ToNot(BeEmpty(), "Invalid argument. input.InfrastructureProviders can't be empty when calling InitManagementClusterAndWatchControllerLogs")
+	Expect(os.MkdirAll(input.LogFolder, 0755)).To(Succeed(), "Invalid argument. input.LogFolder can't be created for InitManagementClusterAndWatchControllerLogs")
+
+	client := input.ClusterProxy.GetClient()
+	controllersDeployments := framework.GetControllerDeployments(context.TODO(), framework.GetControllerDeploymentsInput{
+		Lister: client,
+	})
+	if len(controllersDeployments) == 0 {
+		Init(context.TODO(), InitInput{
```
Comment (wfernandes) [inline test/framework/clusterctl/clusterctl_helpers.go:58]: Should we use the `ctx` being passed in?

**Comment by fabriziopandini** [comment]
@wfernandes thanks for the feedback!
comment addressed

**Comment by fabriziopandini** [comment]
@wfernandes @sedefsavas for final lgtm
the job now runs in parallel and we are down to 20m for 3 test, which is not bad vs 40m that capd-e2e takes for 2 only
